### DISCUSSION TITLE
Purpose and Roadmap of linear-algebra

### DISCUSSION THREAD

**Discussion by snunez1** [Lisp-Stat/linear-algebra#5] [General]
In an email conversation, @beberman asked:

> Are we trying to create the library (libraries) for those who want to do numerical computations using Lisp?  Stats in Lisp?  If so would recommend modeling on NumPy and assorted libraries (Pandas, SciPy, Sympy) etc.

and this seems like a good place to discuss.

The easiest answer is 'yes', the ecosystem that currently lives in the Lisp-Stat repository is intended for scientific and numerical computations using Common Lisp.

A slightly longer answer is that Lisp-Stat started as a faster alternative to R and for the most part is modeled on R.  Of course that's a very statistics oriented perspective, and statistics is only a part of scientific computing.  Little was inspired by Python; not because there's no good ideas there, but because the historical contributors to Lisp-Stat don't know Python.

There is a similarity with Python in design philosophy: Independently useful systems.  Lisp-Stat has special-functions, distributions, plotting, data-frame, statistical functions, array-operations, etc, all of which can be independently pulled into Common Lisp projects without hauling in everything (which you cannot do in R).

Linear algebra, for example, is an independent but related area of numerical computation.  Many statistical computations are easier using linear algebra.  The reason it 'lives' here is really just one of convenience.  Perhaps someday linear-algebra will have its own website, organisational repo and ecosystem.  I hope it does.

On the specific libraries mentioned, there are some Common Lisp versions of those.  An especially well developed one is [Maxima](https://maxima.sourceforge.io/), which is similar to Sympy.  Much of the functionality of SciPy and Pandas resides in Lisp-Stat.  For example the statistical routines, [cephes.cl](https://github.com/Lisp-Stat/cephes.cl), use the same underlying C library that SciPy does.

Finally, it's worth mentioning a 'philosophy' of not reinventing the wheel; a lesson learned the hard way.  R, Python and Julia all derive much of their functionality through language bindings to underlying C or C++ libraries.  When I first started Lisp-Stat I thought I wanted 'pure lisp' all the way through.  That's possible! Consider the [special-functions](https://lisp-stat.dev/docs/resources/special-functions/) system.  Common lisp implementations of these functions can beat, on some intervals, the Boost math library for accuracy!  But at what cost?  A very high development cost, that's what.  In hindsight, special-functions was a useful experiment to demonstrate that common lisp _can_ be applied in that domain, but it's not practical.  Thus the cephes.cl bindings to netlib.  That's a long way of saying the current philosophy is to bind to existing libraries where possible and only develop new functionality when and where required (and in Common Lisp).  And that is a good segue into another discussion on merging [LLA](https://github.com/Lisp-Stat/lla) (BLAS/LAPACK) into linear-algebra and adding a CUDA backend.

**Comment by beberman** [comment]
Good discussion on the purpose and requirements for this library.

Some requirements that might be worth considering. In no particular order:

1. Provide a library interface for multi-dimensional arrays that's easy to use, and provides a consistent interface across underlying implementations BLAS, CUDA, other C-libraries
2.  Make it simple and easy to use for basic linear algebra tasks - creation / manipulation of vectors and matrices, but don't go overboard in terms of the basic implementation in purse lisp. I see no reason for example to write an SVD decomposition in lisp.  If I as a user only want to create some matrices and vectors and multiple them I shouldn't have to install BLAS on my windows machine and get it to work.
3. Make it scalable - should be memory and cpu efficient So memory efficiencies means specialized matrix forms should not have a pile of 0's in the implementation. i.e. a diagonal matrix only stores the diagonal.
4.CPU efficient means take advantage of specialized matrix structures - don't multiple a bunch of 0s if you know they are 0
5. Be parsimonious - i.e. write as little code as needed so its easy to maintain with a few people.  So leverage the built in array and vector capabilities of common lisp
6. Be multiple LISP - should run on SBCL, other lisps.
7. Easy to learn and doc - be as close to NumPy as possible
8. Install from quicklisp
9. Compatible with numerical-utilities (or built on?) and lisp-stat should build on the interface
10. Support multiple dimensional arrays as easily as 2D arrays - 3 and 4 dimensions are often needed in ML
11. Be scalable  to very large sparse matrices.  Actual algorithms at large scale may require BLAS
Thoughts?

**Reply by snunez1** [reply]
I like these as guiding principles.  I'll elaborate/comment on a few of them:

**3** It sounds like you already know that the most common matrix types that require considerations for efficiency already exists in `numerical-utilities`, but it's worth highlighting for any readers new to the discussions/systems.
**5** We should probably be practical about this.  The principle I use in Lisp-Stat is: "Try not to do anything implementation specific unless you have to".  With limited resources, the core developers are realistically only going to be able to support one implementation, and the hope is the community will support others.  This has partially turned out to be true, with contributions for Linux and MacOS coming from non-core contributors.  Some lisps, e.g. CCL (sadly) have some problems in their floating point implementations that, unless fixed, mean they're never going to be good candidates for numerical computing.
**6** Although I appreciate this sentiment, if we look at [numcl](https://github.com/numcl/numcl)'s attempt at this what we got was very unidiomatic lisp, and I found that even with the good Numpy documentation it was a struggle to understand given its 'strange ways' of doing things.

Although slightly tangential, one further comment on documentation: I think this is a critical success factor, and we'll have to accept the fact that we're going to have to write this ourselves, with a good bit of copy & paste (and attribution) from the Python docs.  One hurdle here is the lack of good documentation systems for CL.  I have been looking at this problem for a couple of years now and think I finally have a roadmap for a documentation system using existing libraries.  So whilst all the bits & pieces are there, and a design exists for tying them together into a professional level (literally, Microsoft or VMware quality) documentation system, it's still non-trivial.  I estimate about 2 man-months of effort. :-(

On a design philosophy note, here's a brief explanation of why I think _the system should choose the implementation_:

> Whilst working with matrices over the last few weeks, I had the idea that I'd try to work in some of the `linear-algebra` routines.  At first I explored the original idea of separating function implementations by package space, i.e. linalg:foo vs lla:foo.  This works, but maintaining two separate systems and keeping them in sync is a bit unwieldy.  Then when I needed norms, I implemented them in `linear-algebra`, however they proved to be inadequate, so thought of combining `linear-algebra` and `lla` into a single system and use naming conventions to indicate the implementation, e.g. `norm` is the lisp version and `norm*` the BLAS based version.  This also proved to be unsatisfactory to use and maintain (different parameters for one). At that point I started to wonder 'why should I care?', eventually coming to the conclusion that users in Python and R don't care, and neither should Common Lisp users. We're just making more work for ourselves with this, and the system should mix & match the implementation according to the best resources available and abstract away the implementation details.

This [post on scicomp](https://scicomp.stackexchange.com/questions/3214/what-are-the-fastest-available-implementations-of-blas-lapack-or-other-linear-al) (stackoverflow) describes a way forward pretty well.  Summarised:

- Base the system on [MAGMA ](https://icl.utk.edu/magma/)to abstract away GPU/CPU computations
- Use [Intel MKL](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html) for non-matrix and specialised operations and use cases, like [OneDNN](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onednn.html)
- Keep `linear-algebra` for newbies and tire-kickers that won't/can't setup shared libraries (though that's also a problem I think should be solved sooner rather than later by forking the SBCL installer and putting all our shared libraries in SYS:LIB).

I haven't thought through what to do with `lla`.  It uses the FORTRAN interface, allegedly faster than the CFFI one.  Ideally we could extend `lla` rather than replace it.

Well, that's one vision.  I guess whilst I'm on the topic, I'll add one 'wish' for numerical computing in Common Lisp: [RStudio](https://en.wikipedia.org/wiki/RStudio).  RStudio uses `reticulate` (a [R Interface to Python](https://rstudio.github.io/reticulate/)).  This bridge is similar to the one used in various Common Lisp Jupyter Notebooks.  With that we could tap into the entire R ecosystem, and have an IDE for numerical computing.  It might be worth looking into RCL (Carlos Ungil -- cliki.net down at the time of this writing) or [RCGL](https://github.com/blindglobe/rclg), now abandoned libraries for interfacing to R.  The RStudio license is problematic for commercial usage, so some organisational hurdles would have to be overcome to get this into corporate environments, but I think these are solvable (the RStudio organisation would have to be convinced to support the CL backend because _they_ can offer non-restricted (i.e. non-GNU) licenses, but no one else can).

**Reply by beberman** [reply]
Basing the system on MAGMA and Intel MKL does look like a very good idea.  Can we put a simple installer in place without forking SBCL?  

i.e. possible to make things relatively easy for users so they can install say a merge of lla / linear-algebra into one library and then add those packages for additional speed functionality?

Going back to the requirements list now re-organized:
**Scalability and Tensors**

- Provide a library interface for multi-dimensional arrays that's easy to use, and provides a consistent interface across underlying implementations BLAS, CUDA, other C-libraries
- Make it scalable - should be memory and cpu efficient So memory efficiencies means specialized matrix forms should not have a pile of 0's in the implementation. i.e. a diagonal matrix only stores the diagonal.
- CPU efficient means take advantage of specialized matrix structures - don't multiple a bunch of 0s if you know they are 0
Be parsimonious - i.e. write as little code as needed so its easy to maintain with a few people. So leverage the built in array and vector capabilities of common lisp
- Support multiple dimensional arrays

Current concept is to base on MAGMA/ MKL with an installer that adds these.  The LISP package is a wrapper around these libraries automatically getting scalability and breadth of capability.

**Initial Easy of Use**
- Make it simple and easy to use for basic linear algebra tasks - creation / manipulation of vectors and matrices, but don't go overboard in terms of the basic implementation in purse lisp. 
- Install from Quicklisp
- Get started easily without extract packages?

Installer from SBCL fork or some extended install package?

**Doc and Alignment with NumPy**
- NumPy is to non-LISP so hard for LISP users to understand.
- Lift doc where possible from NumPy and write our own doc.

 **Portable** 
- Can only support SBCL with a few library developers.  Other's need to join community to support other LISPs.

Compatible with numerical-utilities (or built on?) and lisp-stat should build on the interface
Support multiple dimensional arrays as easily as 2D arrays - 3 and 4 dimensions are often needed in ML
Be scalable to very large sparse matrices. Actual algorithms at large scale may require BLAS
Thoughts?

 **Numerical-Utilities** 
- Not clear to me how this should be used/merged in.  It supports Upper, Lower, Diagonal, Hermitian, and regular matrices.  This is a god initial list.   Perhaps Numerical-Utilities should deal with numbers, functions on numbers, polynomials (using lists), root finding etc. So computing log, exp, log Normal (to high precision).  Given a restructuring approach I would probably keep Num-Utils to numerical types and functions on those types and move the solvers to a Num-Solve package.
- Overtime likely additional matrix types will be requested - Unary, Positive Definite, Tri-diagonal, Symmetric (non complex)

**Reply by digikar99** [reply]
I think one important factor to decide the implementation details is whether we want to provide a numerical computing framework for the developer of an end-user application, or for the framework extension developer. What I mean is that basic common lisp using generic functions and macros is sufficient for the first case. But they make writing efficient polymorphic extensions hard if not impossible. 

I have been hacking [numericals](https://github.com/digikar99/numericals) for a while. My goals have been to make efficient polymorphic extensions possible. Unfortunately, I have come to concur with [u/stylewarning's comment on reddit](https://www.reddit.com/r/lisp/comments/16xws4e/comment/k35cxty/?utm_source=share&utm_medium=web2x&context=3). Developing a numerical computing framework that allows the framework extension developers to write efficient polymorphic extensions requires an entire DSL in common lisp. Something similar to julia, although my personal preference seems to be towards ML-like systems. I think coalton is a good candidate for this, but some work is still required for clean [parametric array types in coalton](https://github.com/coalton-lang/coalton/discussions/549).

Such a DSL would address points 1, 3, 9 and 10.

About points 6 on documentation: I recently stumbled upon Eigen/C++. And I'm head over heels about its documentation - https://eigen.tuxfamily.org/dox/. Scientific computing or its use aren't my core areas of interest, so numerical optimization and correctness have always seemed a bit mysterious to me. But I found Eigen's documentation pleasant even with my background; besides documenting the API, they also have very useful pages on guiding users on which algorithms to use for which tasks.

Points 2, 4, 5, 7 are indeed something I can agree upon. 

I find point 8 a bit problematic because of the above discussion about efficient polymorphic extensions requiring a DSL. That said, we should indeed try to make a numerical-utilities or lisp-stat API clone over whatever works out best. So existing lisp-stat users should be able to use by merely changing the `:depends-on` and/or `:use` and/or `:local-nicknames`.

**Reply by beberman** [reply]
Thanks that is an interesting set of comments.  In terms of polymorphic extensions - If I under you - this issue arises from wanting the types within the arrays to have any numerically type. The Coalton suggestion allows this because Coalton supports the type hierarchy.  For numerical statistical computing not sure this is really an issue. Everyone either uses float or double and if we are going to push down into MAGMA which then pushes down into CUDA, LAPACK or BLAS there is an restriction within that C code on the allowed types.

Am I missing something here?

**Reply by digikar99** [reply]
I'd slightly rephrase and elaborate to make sure we are on the same page:

> In terms of polymorphic extensions - If I under you - this issue arises from wanting the types within the arrays to have any numerically type.

In terms of polymorphic extensions - If I under you - this issue arises from wanting a single implementation of an algorithm to work on single-float and double-float (and possibly integers, rationals, complex single-float, complex double-float, short-float, long-float). The part I want to emphasize is that the extension developer should only be required to write a single implementation. This implementation would be polymorphic. The extension user can, then, compile the polymorphic implementation into a monomorphic version suitable to the user's own compute needs. Standard common lisp does not provide any facilities beyond this, beyond what SBCL already does.

> For numerical statistical computing not sure this is really an issue. 

If the set of statistical and/or linear-algebra operations we want to support is a closed set, then the issue of polymorphic extensions is irrelevant. And while that would be a commendable goal, I don't think that the diverse needs of common lisp users would be solved by a closed set of computing operations.

> Everyone either uses float or double and if we are going to push down into MAGMA which then pushes down into CUDA, LAPACK or BLAS there is an restriction within that C code on the allowed types.

As far as I understand magma, the magma user has to decide which particular floating point numbers to use. And once they make that choice, they are stuck with that until they rewrite their code, for example, to replace `f` with `d`. I'd want to aim for a higher level abstraction that makes it easy to choose between different floating point numbers without such a rewrite. 

---

Let me try to illustrate with an example. Suppose that an underlying  library provides `flen dlen fsum dsum` for finding the length and sum of single-float vectors and double-float vectors respectively. Now I - as an extension developer - wish to implement a a `mean` function. 

There are a couple of ways to do it in standard common lisp. The most attractive I find is:

```lisp
(defun sum (vector)
  (etypecase vector
    ((vector single-float)
     (fsum vector))
    ((vector double-float)
     (dsum vector))))

(defun len (vector)
  (etypecase vector
    ((vector single-float)
     (flen vector))
    ((vector double-float)
     (dlen vector))))

(defun mean (vector)
  (/ (sum vector)
     (len vector)))
```

The main downside this has are (i) `sum` and `len` are no longer open to being extended to all the different kinds of integer arrays, or even complex float arrays. (ii) A high level operation such as `mean` is going to perform tons of runtime dispatch on the array type. In the particular case of `mean`, such runtime dispatch would be necessary for `/`, `sum` as well as `len`. This is going to result in slow code quite quickly. 

An alternative is:

```lisp
(defun mean (vector)
  (etypecase vector
    ((vector single-float)
     (/ (fsum vector)
        (flen vector)))
    ((vector double-float)
     (/ (dsum vector)
        (dlen vector)))))
```

With this implementation of `mean`, a single runtime dispatch on the array element type is sufficient. Unfortunately, this, again, suffers from problem (i) - `mean` is not extensible to other floating point formats or complex or integers.

Fortunately, in this trivial case, SBCL performs the dispatching pretty well. It can eliminate runtime dispatch when the functions are inlined. SBCL's dispatching is restricted to builtin array types. Builtin arrays have no notion of layouts, nor of diagonal arrays, or sparse arrays, or hermitian, or being on the CPU vs. GPU. All these features require new user defined array types, and at that point, the SBCL-like approach blows up. Something like coalton becomes necessary. (numcl and julia's might be an okay approach too, but I'm aversive to their JAOT compilation for development purposes.)

Something like coalton can enable us to write:

```lisp
;;; Well, this is not coalton, but I'm hoping this would convey the idea.
(define-function-instance sum ((vector (vector single-float)))
  (fsum vector))
(define-function-instance sum ((vector (vector double-float)))
  (dsum vector))

(define-function-instance len ((vector (vector single-float)))
  (flen vector))
(define-function-instance len ((vector (vector double-float)))
  (dlen vector))

(define-polymorphic-function mean (vector)
  (/ (sum vector)
     (len vector)))

(defun mean-user-1 (vector)
  (declare (type (vector single-float) vector)
           (optimize speed))
  (mean vector))

(defun mean-user-2 (vector)
  (declare (type (vector double-float) vector)
           (optimize speed))
  (mean vector))

```

Here, (i) `sum` and `len` are extensible to other array types. (ii) A single implementation of `mean` is required. This single implementation is compiled to an efficient version specializing on `(vector single-float)` in `mean-user-1` and on `(vector double-float)` in `mean-user-2`. In each of these two functions, no runtime dispatch would be necessary.

--- 

Just to reiterate, all this is a worry only if polymorphic extensions are a concern. If they are not a concern, and there is a good chunk of users who'd be happy with a closed set of statistical operations, then, indeed, a statistical computing framework suited for those users' needs can be built out of standard common lisp.

**Comment by snunez1** [comment]
Well, I think we have a way forward.  Probably the best start is with the installer, as that fixes several other problems in the ecosystem wrt libraries.  Since we do have some useful things in `linear-algebra`, perhaps we can preserve them by migrating those useful functions into `lla` using the `*` alternative function convention, e.g. `solve*`, and all normally named functions will come from Magma and MKL.

**Reply by beberman** [reply]
Great - let me know what tickets you want worked.
